File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque

N = int(input())
square = [list(map(int,input())) for _ in range(N)]
dx = [0,0,1,-1]
dy = [1,-1,0,0]
dangi = 0
dangi_list = []
vis=[[0]*N for _ in range(N)]
def bfs(q):
    temp=1
    while q:
        x,y=q.popleft()
        for dir in range(4):
            nx,ny=x+dx[dir],y+dy[dir]
            if nx<0 or ny<0 or nx>=N or ny>=N:
                continue
            if vis[nx][ny] or not square[nx][ny]:
                continue
            temp+=1
            vis[nx][ny]=1
            q.append((nx,ny))
    dangi_list.append(temp)
for i in range(N):
    for j in range(N):
        if vis[i][j] or not square[i][j]:
            continue
        vis[i][j]=1
        queue=deque()
        queue.append((i,j))
        bfs(queue)
        dangi+=1
print(dangi)
dangi_list.sort()
for i in dangi_list:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;

public class Main {

  private static int[] dx = {0, 0, 1, -1};
  private static int[] dy = {1, -1, 0, 0};
  private static int N;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    N = Integer.parseInt(br.readLine());
    int[][] board = new int[N][N];
    for (int i = 0; i < N; i++) {
      String line = br.readLine();
      for (int j = 0; j < N; j++) {
        board[i][j] = line.charAt(j) - '0';
      }
    }

    List<Integer> order = new ArrayList<>();
    boolean[][] vis = new boolean[N][N];
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < N; j++) {
        if (board[i][j] == 0 || vis[i][j]) {
          continue;
        }
        int count = 1;
        vis[i][j] = true;
        Deque<Point> queue = new ArrayDeque<>();
        queue.offer(new Point(i, j));
        while (!queue.isEmpty()) {
          Point point = queue.poll();
          for (int dir = 0; dir < 4; dir++) {
            int nx = point.x + dx[dir];
            int ny = point.y + dy[dir];
            if (nx < 0 || nx >= N || ny < 0 || ny >= N) {
              continue;
            }
            if (vis[nx][ny] || board[nx][ny] == 0) {
              continue;
            }
            vis[nx][ny] = true;
            queue.offer(new Point(nx, ny));
            count++;
          }
        }
        order.add(count);
      }
    }
    Collections.sort(order);
    StringBuilder sb = new StringBuilder();
    sb.append(order.size()).append("\n");
    for (int number : order) {
      sb.append(number).append("\n");
    }
    System.out.println(sb);
  }

  public static class Point {

    int x;
    int y;

    public Point(int x, int y) {
      this.x = x;
      this.y = y;
    }
  }
}
